### ISSUE TITLE
Golang Modules integration

### ISSUE THREAD

**Issue by ChronosX88** [gotk3/gotk3#501]
We can integrate Go Modules versioning to this library and drop build tags. This will be easier and safer.

**Comment by founderio** [comment]
So instead of the build tag one would import/target a different version according to the gtk version?
An idea worth exploring.

Any suggestion how to do this ensuring that
* Switching the gtk version is seamless (should not be required to update import paths when using a newer version of gtk)
* Code is not duplicated across different versions of gtk

Please keep in mind that
* This project targets all gtk3 versions at the same time - new stuff is just disabled depending on what you compile with
* New functions are added to existing wrapper types across several `*since_3_*.go` files.
* There are some special cases in `*until_3_*.go` where the calls were actually replaced with different calls between versions

Disclaimer: I'm just voicing an opinion, I'm not the maintainer here.

**Comment by andre-hub** [comment]
for Go Modules need all users the golang version 1.11:
https://github.com/golang/go/wiki/Modules

**Comment by ChronosX88** [comment]
@founderio Go Modules uses dependency manifest file. And library versions are pinned to Git tags/commits, so just for each GTK version we should create tag. End developer should take care on which version he is using (specified in dependencies manifest). That's all!

**Comment by founderio** [comment]
But that means we'd have separate development streams, one for each supported GTK version.
So maintenance of 3.{6,8,10,12,14,16,18,20,22,24} in parallel branches, then re-configuring the tags each time anything changes. Not a good solution.

**Comment by andre-hub** [comment]
I agree, with two branches this may still be possible. Which is the case with normal libraries or applications, but here we try to span many variants.

**Comment by ChronosX88** [comment]
@founderio but... Minor versions?

**Comment by founderio** [comment]
Point is, there are several GTK versions out there, so we need to support
them in parallel.
Gotk release cycles do not coincide with gtk release cycles, bindings are
added when needed, not when they are added to gtk.
The bindings are not written for each minor version separately, but rather
all at once.

Essentially this means that adding a new binding for something introduced
in gtk 3.10 means we would need to adapt the branches for 12 14 16 18 20 22
24 as well.
That would be a maintenance hell.

A more feasible solution I can imagine would be to auto-generate branches
based on the build tag info, but who knows how well that would work.

**Comment by ChronosX88** [comment]
@founderio Hm... I agree with you. So you mean that my idea isn't viable? If it's true, then we can just close this issue. Just when I wrote this proposal, I didn’t quite know how your maintenance process is going.

**Comment by andre-hub** [comment]
The general support of go modules should be made available for the respective subsystems. 
So there is a possibility to make the executable file smaller again, if the developer only uses the submodules he needs.
But this will be a bigger change.
